File: androidApp/src/main/java/com/lb/htwktimetable/android/ui/timetable/TimetableContent.kt
package com.lb.htwktimetable.android.ui.timetable

import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.*
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.*
import com.lb.htwktimetable.android.ui.timetable.state.TimetableState
import com.lb.shared.utils.calendar.*

private val hours = buildList {
    for (i in 7..22) {
        if (i < 10) add("0$i:00")
        else add("$i:00")
    }
}

private val hourHeight = 64.dp

private fun calculatePosition(startTime: LocalTimeUtil): Double {
    val hour = startTime.hour - 7
    val min = startTime.minute.toDouble() / 60
    return (hour + min)
}

@Composable
fun TimetableState.TimetableContent() {
    Column(modifier = Modifier.fillMaxSize()) {
        Header(
            modifier = Modifier
                .fillMaxWidth()
                .padding(bottom = 16.dp),
        )
        Box(modifier = Modifier.wrapContentSize().verticalScroll(rememberScrollState())) {
            HourSlots()
            WekkdayColumns()
        }
    }
}

@Composable
private fun HourSlots() {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier
            .fillMaxWidth()
    ) {
        repeat(hours.size) { hour ->
            Box(modifier = Modifier.height(hourHeight)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Box(
                        contentAlignment = Alignment.Center,
                        modifier = Modifier.width(48.dp)
                    ) {
                        Text(
                            text = hours[hour],
                            fontSize = 12.sp
                        )
                    }
                    Divider(modifier = Modifier.height(1.dp))
                }
            }
        }
    }
}

@Composable
private fun TimetableState.WekkdayColumns() {
    LazyRow(
        modifier = Modifier
            .padding(start = 48.dp)
            .fillMaxSize()
    ) {
        items(weekDays) { day ->
            Row(
                modifier = Modifier
                    .fillParentMaxWidth(1 / 7f)
                    .fillMaxHeight()
            ) {
                Box(
                    modifier = Modifier
                        .width(1.dp)
                        .height(hourHeight * 16)
                        .background(Color.LightGray)
                )
                Box {
                    lectures?.forEach {
                        if (it.weeks.contains(week?.calendarWeek.toString()) && it.weekday == day) {
                            LessonCard(
                                lecture = it,
                                modifier = Modifier
                                    .padding(top = calculatePosition(LocalDateExt.stringToLocalTime(it.start)) * hourHeight + 8.dp)
                                    .height(LocalDateExt.calculateHours(it.start, it.end) * hourHeight)
                            )
                        }
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/timetable/TimetableContent.swift
import Foundation
import SwiftUI
import shared

let hourHeight: CGFloat = 52
let weekdays = ["Montag", "Dienstag", "Mittwoch", "Donnerstag", "Freitag", "Samstag", "Sonntag"]

/*
 View for the timetable with its columns and schedules.
 */
struct TimetableContent: View {
    let lectures: [LectureObjectDto]?
    let hours: [String] = (7...22).map { "\($0):00" }
    
    var body: some View {
        ScrollView(.vertical) {
            ZStack {
                // Divider between the hours.
                VStack(alignment: .leading, spacing: hourHeight-1) {
                    ForEach(hours, id: \.self) { hour in
                        Divider()
                            .background(Color.gray)
                            .offset(x: 40, y: -10)
                            .frame(height: 1)
                    }
                }
                // Timetable.
                HStack(alignment: .top) {
                    HourSlots(hours: hours)
                    WeekdayColumns(lectures: lectures)
                    
                }
                .frame(minHeight: UIScreen.main.bounds.height)
            }
            
        }
    }
}

/*
 View for the hours on left side of the screen.
 */
struct HourSlots: View {
    let hours: [String]
    
    var body: some View {
        VStack(alignment: .center, spacing: 0) {
            ForEach(hours, id: \.self) { hour in
                Text(hour).font(.system(size: 12))
                    .padding(.leading, 4)
                    .frame(height: hourHeight)
            }
        }
    }
}

/*
 View for the days with its schedules.
 */
struct WeekdayColumns: View {
    let lectures: [LectureObjectDto]?
    
    var body: some View {
        HStack(spacing: 0) {
            ForEach(weekdays.indices, id: \.self) { index in
                ZStack(alignment: .top) {
                    // Background
                    Rectangle()
                        .fill(Color.clear)
                        .frame(maxWidth: .infinity)
                        .border(Color.gray, width: 0.5)
                    
                    // LessonCards for the day.
                    ZStack(alignment: .top) {
                        ForEach(lectures?.filter { $0.weekday == weekdays[index] } ?? [], id: \.self) { lecture in
                            LessonCard(lecture: lecture)
                                .offset(y: calculateCardOffset(startTime: lecture.start))
                        }
                    }
                }
            }
        }
    }
}

/*
 Calculates the vertical position of a lecture depending on its starttime.
 */
func calculateCardOffset(startTime: String) -> CGFloat {
    let start = LocalDateExt().stringToLocalTime(timeString: startTime)
    let hour = CGFloat(start.hour - 7)
    let min = CGFloat(start.minute) / 60
    return hour * hourHeight + min * hourHeight + 26
}

/*
 Calculates height of a lecture card with Calendar Util function depending on start and endtime.
 */
func calculateCardHeight(start: String, end: String) -> CGFloat {
    return LocalDateExt().calculateHours(start: start, end: end) * hourHeight
}

